operators/src/pro/cache/cache_chunks.rs

```diff
@@ -0,0 +1,398 @@
+            }
+            LandingZoneQueryFeatures::MultiLineString(v) => {
+                v.iter().map(FeatureCollectionInfos::byte_size).sum()
+            }
+            LandingZoneQueryFeatures::MultiPolygon(v) => {
+                v.iter().map(FeatureCollectionInfos::byte_size).sum()
+            }
+        }
+    }
+}
+
+impl From<LandingZoneQueryFeatures> for CachedFeatures {
+    fn from(value: LandingZoneQueryFeatures) -> Self {
+        match value {
+            LandingZoneQueryFeatures::NoGeometry(v) => CachedFeatures::NoGeometry(Arc::new(v)),
+            LandingZoneQueryFeatures::MultiPoint(v) => CachedFeatures::MultiPoint(Arc::new(v)),
+            LandingZoneQueryFeatures::MultiLineString(v) => {
+                CachedFeatures::MultiLineString(Arc::new(v))
+            }
+            LandingZoneQueryFeatures::MultiPolygon(v) => CachedFeatures::MultiPolygon(Arc::new(v)),
+        }
+    }
+}
+
+impl CacheElementsContainerInfos<VectorQueryRectangle> for CachedFeatures {
+    fn is_expired(&self) -> bool {
+        self.is_expired()
+    }
+}
+
+impl<G> CacheElementsContainer<VectorQueryRectangle, FeatureCollection<G>> for CachedFeatures
+where
+    G: CacheElementSubType<CacheElementType = FeatureCollection<G>> + Geometry + ArrowTyped,
+    FeatureCollection<G>: CacheElementHitCheck,
+{
+    type ResultStream = CacheChunkStream<G>;
+
+    fn result_stream(&self, query: &VectorQueryRectangle) -> Option<CacheChunkStream<G>> {
+        G::result_stream(self, query)
+    }
+}
+
+impl<G> LandingZoneElementsContainer<FeatureCollection<G>> for LandingZoneQueryFeatures
+where
+    G: CacheElementSubType<CacheElementType = FeatureCollection<G>> + Geometry + ArrowTyped,
+    FeatureCollection<G>: CacheElementHitCheck,
+{
+    fn insert_element(
+        &mut self,
+        element: FeatureCollection<G>,
+    ) -> Result<(), super::error::CacheError> {
+        G::insert_element_into_landing_zone(self, element)
+    }
+
+    fn create_empty() -> Self {
+        G::create_empty_landing_zone()
+    }
+}
+
+impl<G> CacheElement for FeatureCollection<G>
+where
+    G: Geometry + ArrowTyped + CacheElementSubType<CacheElementType = Self> + ArrowTyped + Sized,
+    FeatureCollection<G>: CacheElementHitCheck,
+{
+    type Query = VectorQueryRectangle;
+    type LandingZoneContainer = LandingZoneQueryFeatures;
+    type CacheContainer = CachedFeatures;
+    type ResultStream = CacheChunkStream<G>;
+    type CacheElementSubType = G;
+
+    fn cache_hint(&self) -> geoengine_datatypes::primitives::CacheHint {
+        self.cache_hint
+    }
+
+    fn typed_canonical_operator_name(
+        key: crate::engine::CanonicOperatorName,
+    ) -> super::shared_cache::TypedCanonicOperatorName {
+        super::shared_cache::TypedCanonicOperatorName::Vector(key)
+    }
+
+    fn update_stored_query(&self, _query: &mut Self::Query) -> Result<(), CacheError> {
+        Ok(())
+    }
+}
+
+macro_rules! impl_cache_element_subtype_magic {
```

Comment:
what exactly is the magic here?

Reply:
[b8ddf51](https://github.com/geo-engine/geoengine/pull/833/commits/b8ddf51a5004245a510b71fee97e3c5dcd50cc7c)